feat: add installation and run scripts for Windows, Linux, and Docker support

KiringYJ opened the pull request:
This PR adds installation and run scripts for both Windows and Linux environments, making it easier to set up and manage the Urocissa project. While the project already had cross-platform compatibility, these scripts simplify the installation process and provide Docker support for testing and deployment.

## **Key Changes**  

1. **Installation Script (`install-urocissa.mjs`)**:  
   - Automates dependency installation for both Windows and Linux (`Rust`, `Node.js`, and `FFmpeg`).
   - Includes checks for existing installations and prompts for updates.
   - Supports `--debug` and `--update` flags for flexible debugging and updating workflows.

2. **Run Script (`run-urocissa.mjs`)**:  
   - Automates backend startup using `cargo run --release` while maintaining directory navigation.

3. **Dockerfile**:  
   - Sets up a lightweight Ubuntu-based container with all necessary dependencies for running the project.
   - Copies project files and ensures scripts are executable.
   - Simplifies testing in a controlled Linux environment.

## **How to Test**

### **Testing on Windows and Linux**  
1. Run the installation script in Node.js:  
   ```bash
   node install-urocissa.mjs
   ```
2. Start the backend:  
   ```bash
   node run-urocissa.mjs
   ```

### **Testing with Docker**  
1. **Build the Docker Image**:  
   ```bash
   docker build -t urocissa-test .
   ```

2. **Run the Docker Container**:  
   ```bash
   docker run -it -p 4000:4000 urocissa-test
   ```

3. **Access the Application**:  
   Open a browser and navigate to:  
   ```
   http://127.0.0.1:4000
   ```

4. **Clean Up After Testing**:  
   - Stop the container by exiting (`Ctrl+C`) or using `docker ps` and `docker stop`.
   - Remove the container with:
     ```bash
     docker rm <container_id> -f
     ```

## **Checklist**
- [x] Added run scripts for Windows and Linux.
- [x] Added installation scripts for Windows and Linux.
- [x] Verified Docker functionality and application accessibility on `http://127.0.0.1:4000`.

## **Additional Notes**
- These scripts simplify the setup process for the Urocissa project without altering its cross-platform compatibility.
- Docker support enables seamless testing in a Linux environment, even on non-Linux systems like Windows.
- Future enhancements could include CI/CD pipelines for automated builds and tests in Windows, Linux, and Docker environments.

hsa00000 commented:
Hi @KiringYJ,

Thank you for your contributions and the detailed explanations in your pull request! I appreciate the effort you've put into enhancing the project. I have a few questions and concerns that I'd like to clarify before proceeding with the merge:

1. **Change in `package.json` Build Script**
   I noticed that in `gallery-frontend/package.json`, the "build" script was updated to:

   ```json
   "prebuild": "npm ci",
   "build": "vue-tsc --noEmit && vite build"
   ```

   However, I didn't see `prebuild` being utilized in the `.mjs` installation scripts. This seems problematic because if the frontend dependencies are updated, and `npm ci` is not run beforehand, the `node install-urocissa --update` command might fail to build the frontend properly. Could you clarify if `prebuild` should be explicitly invoked or integrated into the update process? Otherwise, this could lead to issues for users.

2. **File Handling in `.gitignore` and `install-urocissa.mjs`**
   Should `ffmpeg.zip` and `rustup-init.exe` be added to `.gitignore`? Also, while `ffmpeg.zip` is deleted after installation in `install-urocissa.mjs`, it seems `rustup-init.exe` is not being removed. Wouldn't it be more consistent and cleaner to delete `rustup-init.exe` as well after installation?

3. **Purpose and Usage of Docker**\
   Could you explain the intended use case for the Docker setup? It’s not entirely clear how this Dockerfile is meant to be used.

   If it’s primarily for testing the `.mjs` scripts, wouldn't it be unnecessary to pre-install `ffmpeg` and `rust` in the Dockerfile since the `.mjs` scripts already handle these installations? Additionally, I noticed there’s no command in the Dockerfile to execute the `.mjs` scripts. Should this be included, or should the Dockerfile be simplified for its specific purpose?

   Moreover, if existing users utilize your Dockerfile, they will end up copying the entire project folder into the Docker container, including backend-stored photos/videos and the frontend's npm dependencies. This doesn't seem ideal, as it may lead to unnecessary bloat.

Looking forward to your clarification. Thanks again for your hard work on improving the project! I'm more than happy to merge the `.mjs` scripts once the issues in points 1 and 2 are resolved, and I truly appreciate your contributions and the effort you've put in!

Best regards,\
hsa00000

KiringYJ commented:
Hi @hsa00000,

Thank you for your thoughtful feedback and for taking the time to review my pull request. I'd like to address each of your points:

1. **Change in `package.json` Build Script**  
   The `prebuild` hook will be automatically executed before the `build` command, so it doesn't need to be explicitly invoked in the `.mjs` scripts. Regarding the use of `npm install` versus `npm ci`, there's been considerable discussion on this topic online. From my understanding, in a CI/CD environment, `npm ci` is generally preferred for maintaining consistency and compatibility, as it strictly adheres to `package-lock.json`.

2. **File Handling in `.gitignore` and `install-urocissa.mjs`**  
   You are correct; I missed deleting `rustup-init.exe` after the installation. This has now been addressed in my latest commit. I've also added `ffmpeg.zip` and `rustup-init.exe` to `.gitignore` to prevent them from being accidentally included in the repository, even though the script ensures these files are cleaned up after use. This addition is mainly for safety.

3. **Purpose and Usage of Docker**  
   The primary reason for using Docker is to provide an isolated environment, ensuring that the installation does not contaminate the user's system settings. Testing Linux environments on Windows is an additional benefit. I chose to pre-install dependencies via the Dockerfile rather than through the `.mjs` script because it simplifies the Docker setup process, and I wanted to avoid writing separate installation logic specifically for Docker. While there is some redundancy in checking for installed dependencies, I find this acceptable to keep things streamlined.

   Regarding your point about no command being present in the Dockerfile to execute the `.mjs` script—initially, the idea was for users to run the container and then manually execute the script in the Bash shell. However, I realize I forgot to mention this in my earlier comments. In the new commit, I have integrated the installation process directly into the Dockerfile, allowing users to run the container and have the backend start up without additional commands.

   Lastly, I appreciate your point about copying the entire project folder into the Docker container. This was indeed due to me missing a `.dockerignore` file, which I've now added in the latest commit. This should prevent unnecessary files, like backend-stored media or frontend dependencies, from being included in the Docker image, reducing bloat.

I believe these changes address the issues you raised, and I appreciate your help in improving the project.

hsa00000 commented:
Thank you for your work on this PR. The `mjs` scripts look good to me. Regarding the Dockerfile, I plan to make some modifications and add features like `volume` support.  

To ensure these changes can be made without affecting the stability of the `main` branch, I’ve decided to merge your PR into a new feature branch (`feature/add-install-run-script`).  

Once the adjustments are complete, I’ll merge everything into the `main` branch. If you have any further suggestions or input on refining the Docker setup, feel free to let me know.  

Thank you for your contribution!
